[ALL TESTS] Update (#5645)
mlflow/mlflow: CHANGELOG.md

# CHANGELOG

## 1.25.0 (2022-04-11)

MLflow 1.25.0 includes several major features and improvements:

Features:

- [Tracking] Introduce a new fluent API `mlflow.last_active_run()` that provides the most recent fluent active run (#5584, @MarkYHZhang)
- [Tracking] Add `experiment_names` argument to the `mlflow.search_runs()` API to support searching runs by experiment names (#5564, @r3stl355)
- [Tracking] Add a `description` parameter to `mlflow.start_run()` (#5534, @dogeplusplus)
- [Tracking] Add `log_every_n_step` parameter to `mlflow.pytorch.autolog()` to control metric logging frequency (#5516, @adamreeve)
- [Tracking] Log `pyspark.ml.param.Params` values as MLflow parameters during PySpark autologging (#5481, @serena-ruan)
- [Tracking] Add support for  `pyspark.ml.Transformer`s to PySpark autologging (#5466, @serena-ruan)
- [Tracking] Add input example and signature autologging for Keras models (#5461, @bali0019)

> - [Models] Add support for logging models to preexisting destination directories (#5572, @akshaya-a)

Reviewer: Nit: Can we move this entry lower in the `[Models]` subsection since it's a relatively small improvement?

Reviewer 2: moved

- [Models] Introduce `mlflow.diviner` flavor for large-scale [time series forecasting](https://databricks-diviner.readthedocs.io/en/latest/?badge=latest) (#5553, @BenWilson2)

> - [Models] Add `mlflow_version` field to MLModel specification (#5515, #5576, @r3stl355)

Reviewer: Nit: Can we move this entry lower in the `[Models]` subsection since it's a relatively small improvement?

Reviewer 2: yep!

- [Models] Add `pyfunc.get_model_dependencies()` API to retrieve reproducible environment specifications for MLflow Models with the pyfunc flavor (#5503, @WeichenXu123)
- [Models] Add `code_paths` argument to all model flavors to support packaging custom module code with MLflow Models (#5448, @stevenchen-db)
- [Models] Support creating custom artifacts when evaluating models with `mlflow.evaluate()` (#5405, #5476 @MarkYHZhang)

> - [Scoring / Projects] Introduce `--env-manager` configuration for specifying environment restoration tools (e.g. `conda`) Deprecate `--no-conda` (#5567, @harupy)

Reviewer: ```suggestion
- [Scoring / Projects] Introduce `--env-manager` configuration for specifying environment restoration tools (e.g. `conda`), and deprecate `--no-conda` (#5567, @harupy)
```

- [Scoring] Support restoring model dependencies in `mlflow.pyfunc.spark_udf()` to ensure accurate predictions (#5487, #5561, @WeichenXu123)

> - [Scoring] Add support for `numpy.ndarray` type inputs to `tensorflow` predict function (#5545, @WeichenXu123)

Reviewer: ```suggestion
- [Scoring] Add support for `numpy.ndarray` type inputs to the TensorFlow pyfunc `predict` function (#5545, @WeichenXu123)
```

- [Scoring] Support deployment of MLflow Models to Sagemaker Serverless (#5610, @matthewmayo)
- [UI] Add MLflow version to header beneath logo (#5504, @adamreeve)

> - [Artifacts] Introduce a `mlflow.artifacts.download_artifacts()` API mirroring the functionality of the `mlflow artifacts download` CLI (#5585, @dbczumar)

Reviewer 3: Maybe it is just me, but when I see `mlflow.artifacts.download_artifacts()`, I get an impression that `download_artifacts` can be called without arugments.

Reviewer: I don't get that impression, but I'll defer to @BenWilson2 here :)

Reviewer 3: btw what does `xxx()` indicate? Does this indicate `xxx` a callable object?

Reviewer 3: Looks like pandas uses this format:

https://pandas.pydata.org/docs/whatsnew/v1.4.1.html

Reviewer: > btw what does `xxx()` indicate? Does it indicate `xxx` is a function?

Yes

Reviewer 3: I see, I'm personally not a big fan of this format because `xxx()` looks like a function *call*, not a function, but I'm fine because other packages also use this format.

Reviewer 2: yeah I've always used the inclusion of () in doc strings to indicate a method or function (callable) and without it as a module name. This isn't to say that it's correct and I have no particular conviction one way or another (so long as it's kept standardized :) ).

- [Artifacts] Introduce environment variables for controlling GCS artifact upload/download chunk size and timeouts (#5438, #5483, @mokrueger)

Bug fixes and documentation updates:

- [Tracking/SQLAlchemy] Create an index on `run_uuid` for PostgreSQL to improve query performance (#5446, @harupy)
- [Tracking] Remove client-side validation of metric, param, tag, and experiment fields (#5593, @BenWilson2)

> - [Artifacts] Add compatibility for reading `ENV` variables for GCS to prevent timeouts (#5483, @sebastian-montero)

Reviewer: Can we remove this line since we already incorporated the associated PR into the feature changelog entry here: https://github.com/mlflow/mlflow/pull/5645/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR28 ?

> - [Projects] Supporting setting the name of the MLflow Run when executing an MLflow Project (#5187, @bramrodenburg)

Reviewer 3: ```suggestion
- [Projects] Support setting the name of the MLflow Run when executing an MLflow Project (#5187, @bramrodenburg)
```

nit: to align with other entries?

Reviewer 2: changed to verb

- [Scoring] Use pandas `split` orientation for DataFrame inputs to SageMaker deployment `predict()` API to preserve column ordering (#5522, @dbczumar)
- [Server-Infra] Fix runs search compatibility bugs with PostgreSQL, MySQL, and MSSQL (#5540, @harupy)

> - [CLI] Fixed a bug in the `mlflow-skinny` client that caused `mlflow --version` to fail (#5573, @BenWilson2)

Reviewer: ```suggestion
- [CLI] Fix a bug in the `mlflow-skinny` client that caused `mlflow --version` to fail (#5573, @BenWilson2)
```

Reviewer 2: changed to present tense

> # CHANGELOG

Reviewer: Nit: Can we remove trailing periods from changelog entries?

Reviewer 2: removed!

## 1.25.0 (2022-04-11)

MLflow 1.25.0 includes several major features and improvements:

Features:

> - [Artifacts] Introduce a new `mlflow.artifacts.download_artifacts()` API that mirrors functionality of CLI `download_artifacts` (#5585, @dbczumar)

Reviewer: Can we move this to the bottom of the features changelog? 

```suggestion
- [Artifacts] Introduce a `mlflow.artifacts.download_artifacts()` API mirroring the functionality of the `mlflow artifacts download` CLI (#5585, @dbczumar)
```

> - [Models] Add support for saving artifact if destination is an empty directory. (#5572, @akshaya-a)

Reviewer: ```suggestion
- [Models] Add support for logging models to preexisting destination directories. (#5572, @akshaya-a)
```

> - [Models] Introduce support for `diviner` as a flavor. (#5553, @BenWilson2)

Reviewer: Can we add a link to Diviner and a short (1 key phrase) explanation of what it is?

Reviewer 2: docs link added

> - [Models] Introduce support for `diviner` as a flavor. (#5553, @BenWilson2)

Reviewer: ```suggestion
- [Models] Introduce `mlflow.diviner` flavor for Diviner models. (#5553, @BenWilson2)
```

> - [Models] Add `mlflow_version` to logged model metadata. (#5515, @r3stl355)

Reviewer: ```suggestion
- [Models] Add `mlflow_version` to logged model metadata. (#5515, #5576, @r3stl355)
```

> - [Models] Add `pyfunc.get_model_dependencies()` API to retrieve reproducible environment configuration for pyfunc flavor. (#5503, @WeichenXu123)

Reviewer: ```suggestion
- [Models] Add `pyfunc.get_model_dependencies()` API to retrieve reproducible environment specifications for MLflow Models with the pyfunc flavor. (#5503, @WeichenXu123)
```

> - [Models] Add input example and signature autologging for `tensorflow.keras`. (#5461, @bali0019)

Reviewer: ```suggestion
- [Tracking] Add input example and signature autologging for Keras models. (#5461, @bali0019)
```

> - [Models] Add `code_paths` argument to all model flavors to support logging custom module code for external-dependencies for a model. (#5448, @stevenchen-db)

Reviewer: ```suggestion
- [Models] Add `code_paths` argument to all model flavors to support packaging custom module code with MLflow Models. (#5448, @stevenchen-db)
```

> - [Models / Projects] Introduce `--env-manager` configuration to prepare for support for pyenv & virtualenv. Deprecate `--no-conda` (#5567, @harupy)

Reviewer: Nit: Let's not mention virtualenv until we support it.

```suggestion
- [Models / Projects] Introduce `--env-manager` configuration for specifying environment restoration tools (e.g. `conda`). Deprecate `--no-conda` (#5567, @harupy)
```

> - [Scoring] Introduce `--env-manager` configuration for selection of either local or conda environment for serving. (#5487, @WeichenXu123)

Reviewer: Weichen's https://github.com/mlflow/mlflow/pull/5487 PR adds the argument to mlflow.pyfunc.spark_udf() and should be combined with other changelog entries related to model restoration for mlflow.pyfunc.spark_udf() (will leave a separate comment about this).

> - [Scoring] Introduce `--env-manager` configuration for selection of either local or conda environment for serving. (#5487, @WeichenXu123)

Reviewer: ```suggestion
- [Scoring] Support restoring model dependencies in `mlflow.pyfunc.spark_udf()` to ensure accurate predictions (#5487, #5561, @WeichenXu123)
```

- [Scoring] Add support for `numpy.ndarray` type inputs to `tensorflow` predict function (#5545, @WeichenXu123)

> - [Scoring] Add support for Docker image deployment to Sagemaker Serverless (#5610, @matthewmayo)

Reviewer: ```suggestion
- [Scoring] Support deployment of MLflow Models to Sagemaker Serverless (#5610, @matthewmayo)
```

> - [Tracking] Introduce a new fluent API `mlflow.last_active_run()` that will return `mlflow.active_run()` when called. (#5584, @MarkYHZhang)

Reviewer: ```suggestion
- [Tracking] Introduce a new fluent API `mlflow.last_active_run()` that provides the most recent fluent active run. (#5584, @MarkYHZhang)
```

> - [Tracking] Introduce a new fluent API `mlflow.last_active_run()` that will return `mlflow.active_run()` when called. (#5584, @MarkYHZhang)

Reviewer: Can we list Tracking updates first, since Tracking is the most widely-used MLflow component?

Reviewer 2: Tracking -> Models -> Scoring -> UI -> Artifacts

> - [Tracking] Add argument `experiment_names` in the fluent `mlflow.search_runs()` API to support run search by experiment names. (#5564, @r3stl355)

Reviewer: ```suggestion
- [Tracking] Add `experiment_names` argument to the `mlflow.search_runs()` API to support searching runs by experiment names. (#5564, @r3stl355)
```

> - [Tracking] Add support for specifying or overriding a run description in the fluent API `start_run()` with an argument. (#5534, @dogeplusplus)

Reviewer: ```suggestion
- [Tracking] Add a ``description`` parameter to ``mlflow.start_run()`` (#5534, @dogeplusplus)
```

> - [Tracking] Add `log_every_n_step` feature to `mlflow.pytorch.autolog()` to log metrics during training for PyTorch. (#5516, @adamreeve)

Reviewer: ```suggestion
- [Tracking] Add `log_every_n_step` parameter to `mlflow.pytorch.autolog()` to control metric logging frequency (#5516, @adamreeve)
```

> - [Tracking] Add support for `pyspark.ml.param.Params` types for `Estimators`. (#5481, @serena-ruan)

Reviewer: ```suggestion
- [Tracking] Log `pyspark.ml.param.Params` values as MLflow parameters for PySpark `Estimators`. (#5481, @serena-ruan)
```

> - [Tracking] Add support for tracking `pyspark.ml.Transformer` for autologging. (#5466, @serena-ruan)

Reviewer: ```suggestion
- [Tracking] Add support for  `pyspark.ml.Transformer`s to PySpark autologging. (#5466, @serena-ruan)
```

> - [Tracking] Add support for custom metrics to support auto detection and loading of user-generated artifacts. (#5405, @MarkYHZhang)

Reviewer: ```suggestion
- [Models] Support creating custom artifacts when evaluating models with `mlflow.evaluate()`. (#5405, @MarkYHZhang)
```

> - [UI] Update UI (#5577, @harupy)

Reviewer 2: @harupy @BenWilson2 Summarize the UI changes from copybara sync

Bug fixes and documentation updates:

> - [Artifacts] Add compatibility for reading `ENV` variables for GCS to prevent timeouts. (#5483, @sebastian-montero)

Reviewer: Can we combine this with https://github.com/mlflow/mlflow/pull/5645/files#r846543176? We can remove this line and accept the suggestion on https://github.com/mlflow/mlflow/pull/5645/files#r846543176.

Reviewer 2: yep!

> - [Artifacts] Fix for EvaluationResult to save artifacts with appropriate file names. (#5476, @MarkYHZhang)

Reviewer: Can we include this in https://github.com/mlflow/mlflow/pull/5645/files#r846541823 instead of calling it out as a bug fix here?

Reviewer 2: yep

> - [Artifacts] Add `ENV` variables for GCS to control upload/download chunk size and timeouts (#5438, @mokrueger)

Reviewer: This is a feature

```suggestion
- [Artifacts] Introduce environment variables for controlling GCS artifact upload/download chunk size and timeouts (#5438, #5483, @mokrueger)
```

> - [Build] Fix mlflow-skinny version stack frame inspection failure (#5573, @BenWilson2)

Reviewer: Can we focus on the user-facing impact here? I.e fixed a bug in the MLflow skinny client that caused `mlflow --version` to fail. I think `[CLI]` is a good tag here.

Reviewer 2: nice. changed

> - [Docs] Update example for deployment to AzureML and guidance on using the officially supported Microsoft plugin. (#5491, @santiagxf)

Reviewer: Can we list docs updates at the bottom of this changelog section?

```suggestion
- [Docs] Update guidance and examples for model deployment to AzureML to recommend using the `mlflow-azureml` package (#5491, @santiagxf)
```

> - [Models] Add mlflow_version to Java model spec to fix MLeap serving issue (#5576, @dbczumar)

Reviewer: Can we include the PR for this in the feature entry for including `mlflow_version` in the MLflow Model spec? I.e. remove this line and adopt https://github.com/mlflow/mlflow/pull/5645/files#r846544456

> - [Projects] Set the `run_name` argument within an MLproject. (#5187, @bramrodenburg)

Reviewer: This is a feature

```suggestion
- [Projects] Supporting setting the name of the MLflow Run when executing an MLflow Project (#5187, @bramrodenburg)
```

> - [Scoring] Use `split` pandas orientation for sagemaker deployment `predict()` function to preserve column ordering. (#5522, @dbczumar)

Reviewer: ```suggestion
- [Scoring] Use pandas `split` orientation for DataFrame inputs to SageMaker deployment `predict()` API to preserve column ordering (#5522, @dbczumar)
```

> - [Server-Infra] Add backend database testing against PostgreSQL, MySQL, and MSSQL (#5540, @harupy)

Reviewer: Let's focus on the user impact of the PR:

```suggestion
- [Tracking] Fix runs search compatibility bugs with PostgreSQL, MySQL, and MSSQL (#5540, @harupy)
```

Reviewer 2: +1 changed

> - [Server-Infra] Create an index on `run_uuid` for PostgreSQL to improve query performance. (#5446, @harupy)

Reviewer: ```suggestion
- [Tracking/SQLAlchemy] Create an index on `run_uuid` for PostgreSQL to improve query performance (#5446, @harupy)
```

> - [Server-Infra] Add validation checks to cli to prevent server from being started in `--artifacts-only` mode if `--backend-store-uri` is set to non-local (#5571, @BenWilson2)

Reviewer: Are we sure that validating on `--backend-store-uri` is the correct choice here?

Reviewer 2: 100%. deleted

> - [Tracking] Remove client-side validation of metrics, params, tags, and experiment name (#5593, @BenWilson2)

Reviewer: ```suggestion
- [Tracking] Remove client-side validation of metric, param, tag, and experiment fields (#5593, @BenWilson2)
```

> - [UI] Add MLflow version to header beneath logo. (#5504, @adamreeve)

Reviewer: This is a feature

Reviewer 2: moved

# CHANGELOG

## 1.25.0 (2022-04-08)

MLflow 1.25.0 includes several major features and improvements:

Features:

- pyfunc.spark_udf: conda env dir / cache dir isolation and NFS optimization  (#5561, @WeichenXu123)
- Introduce an mlflow.artifacts.download_artifacts() API, equivalent to the preexisting CLI (#5585, @dbczumar)

> - Update UI (#5577, @harupy)

Reviewer: @harupy @BenWilson2 Can we summarize the key UI updates from the copybara import?

Reviewer 3: I just checked https://github.com/mlflow/mlflow/pull/5577 and it doesn't contain any user-facing changes.

Reviewer 3: It contains the changes for cross-experiment run comparison, but I don't think we need to mention it.

Reviewer 3: Maybe jquery removal should be highlighted?

Reviewer: I think we can remove this entry and put it in the "small updates" section.

Reviewer 3: No, it's not. Putting in the small updates section sounds good

Reviewer 2: k will do!
